baseapp/abci.go

```diff
@@ -249,8 +250,19 @@ func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBloc
 // Ref: https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-060-abci-1.0.md
 // Ref: https://github.com/tendermint/tendermint/blob/main/spec/abci/abci%2B%2B_basic_concepts.md
 func (app *BaseApp) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
-	// TODO: Implement.
-	return abci.ResponsePrepareProposal{Txs: req.Txs}
+	memTxs, selectErr := app.mempool.Select(req.Txs, req.MaxTxBytes)
+	if selectErr != nil {
+		panic(selectErr)
+	}
+	var txs [][]byte
+	for _, memTx := range memTxs {
```

Comment:
```suggestion
	}

	txs := make([][]byte, len(memTxs))
	for i, memTx := range memTxs {
```

Since we know the size of `memTxs`, we can pre-allocate.

Reply:
The expression

```go
txs := make([][]byte, len(memTxs))
```

creates a slice of length `len(memTxs)` where each element is the zero value of the slice type, i.e. a nil slice of []byte. That means `append` would add to the _end_ of the slice, and this block of code would produce a `txs` of length `2 * len(memTxs)` where the first half would be nil slices.

If you're going to use `append` then you want

```go
txs := make([][]byte, 0, len(memTxs))
```

or you can keep the current `txs` declaration as long as you do assignment as

```go
txs[i] = bz
```

but there's no real reason to pre-allocate the slice here, as the slice element type `[]byte` is a 3-tuple (very tiny), the practical maximum cardinality of this thing is O(100) or maybe O(1k) (slice allocation of this size is statistically zero in the context of allocations in the the SDK), and this method is only called once every block interval (which is O(seconds) and therefore effectively never).